### PR TITLE
Allow benchmark specific flags

### PR DESCRIPTION
**Pull request by ehankland**
```
./pkb.py --config_override="benchmarks=[{iperf: {flags: {iperf_sending_thread_count: 2}}}, {iperf: {flags: {iperf_sending_thread_count: 4}}}]" --parallelism=2
.....
.....
.....
-------------------------PerfKitBenchmarker Results Summary-------------------------
IPERF:
  receiving_machine_type="n1-standard-1" receiving_zone="us-central1-a" runtime_in_seconds="60" sending_machine_type="n1-standard-1" sending_zone="us-central1-a"
  Throughput                         1919.000000 Mbits/sec                      (ip_type="external" sending_thread_count="4")
  Throughput                         1993.000000 Mbits/sec                      (ip_type="internal" sending_thread_count="4")
  Throughput                         1955.000000 Mbits/sec                      (ip_type="external" sending_thread_count="4")
  Throughput                         1992.000000 Mbits/sec                      (ip_type="internal" sending_thread_count="4")
  Throughput                         1909.000000 Mbits/sec                      (ip_type="external" sending_thread_count="2")
  Throughput                         1997.000000 Mbits/sec                      (ip_type="internal" sending_thread_count="2")
  Throughput                         1758.000000 Mbits/sec                      (ip_type="external" sending_thread_count="2")
  Throughput                         1997.000000 Mbits/sec                      (ip_type="internal" sending_thread_count="2")
  End to End Runtime                  496.858173 seconds                       
  End to End Runtime                  518.701941 seconds                       

-------------------------
```

### CODE REVIEW

**Comment by ehankland** [comment]
@cmccoy Would you mind taking a look?

File: perfkitbenchmarker/benchmark_spec.py [hunk 1]
```diff
@@ -195,6 +196,11 @@ def __init__(self, benchmark_config, benchmark_uid):
     self.deleted = False
     self.file_name = os.path.join(vm_util.GetTempDir(), self.uid)
     self.always_call_cleanup = False
+    self.FLAGS = None
+
+  def DoFlagsMerge(self):
+    """Merge flags from the config with the global flags."""
+    self.FLAGS = configs.GetMergedFlags(self.config)
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: Consider making `self.FLAGS` a lazily initialized property.

File: perfkitbenchmarker/benchmark_spec.py [hunk 2]
```diff
@@ -387,8 +393,10 @@ def DeleteVm(self, vm):
 
   def PickleSpec(self):
     """Pickles the spec so that it can be unpickled on a subsequent run."""
+    flags, self.FLAGS = self.FLAGS, None
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: Why is this needed?

File: perfkitbenchmarker/vm_util.py [hunk 3]
```diff
@@ -90,6 +88,33 @@ class IpAddressSubset(object):
                   'external IP only (EXTERNAL) or internal IP only (INTERNAL)')
 
 
+class FakeFlagValues(object):
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: This doesn't seem like a great place for this class. Maybe introduce a module for current context?

Also (nit): How is this Fake? Maybe `FlagValuesProxy`?

File: perfkitbenchmarker/vm_util.py [hunk 4]
```diff
@@ -90,6 +88,33 @@ class IpAddressSubset(object):
...
+  """Class which provides the same interface as FlagValues."""
+
+  def _GetFlagValues(self):
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Nit: this is lightweight enough to be a property.

File: perfkitbenchmarker/vm_util.py [hunk 5]
```diff
@@ -179,6 +204,28 @@ def GetSshOptions(ssh_key_filename):
   return options
 
 
+class _ThreadData(threading.local):
+  def __init__(self):
+    self.benchmark_spec = None
+
+
+thread_local = _ThreadData()
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Private?

**Comment by cmccoy** [comment]
This could use a README update, and a note in #525.

File: perfkitbenchmarker/pkb.py [hunk 6]
```diff
@@ -280,10 +280,13 @@ def RunBenchmark(benchmark, collector, sequence_number, total_benchmarks,
           # a reference to the spec in order to delete it in the "finally"
           # section below.
           spec = benchmark_spec.BenchmarkSpec(benchmark_config, benchmark_uid)
+          vm_util.SetThreadBenchmarkSpec(spec)
+          spec.DoFlagsMerge()
           spec.ConstructVirtualMachines()
           DoPreparePhase(benchmark, benchmark_name, spec, detailed_timer)
         else:
           spec = benchmark_spec.BenchmarkSpec.GetSpecFromFile(benchmark_uid)
+          vm_util.SetThreadBenchmarkSpec(spec)
```
Comment (cmccoy) [inline perfkitbenchmarker/pkb.py]: Would it be possible to set the current thread benchmark spec during construction?

File: perfkitbenchmarker/vm_util.py [hunk 7]
```diff
@@ -90,6 +88,33 @@ class IpAddressSubset(object):
                   'external IP only (EXTERNAL) or internal IP only (INTERNAL)')
 
 
+class FakeFlagValues(object):
+  """Class which provides the same interface as FlagValues."""
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Please elaborate :)

File: perfkitbenchmarker/vm_util.py [hunk 8]
```diff
@@ -90,6 +88,33 @@ class IpAddressSubset(object):
...
+
+  def _GetFlagValues(self):
+    """Returns the correct FlagValues object for the current thread."""
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Please elaborate.

File: perfkitbenchmarker/benchmark_spec.py (hunk 2 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: Pickle blows up if you try to pickle FlagValues:
PicklingError: Can't pickle <function Checker at 0x7fb410743848>: it's not found as gflags.Checker
Even dill blows up when unpickling:
KeyError: '__flags'

Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: I'll add a comment explaining that

Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: OK. Can you note that here?
Are the FLAGS correctly serialized this way? Or do the same flags need to be specified when the spec is deserialized?

Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: Since the flags specified in the config are part of the config, those get correctly serialized. The command line flags don't get serialized, and have to be re-specified. This is the existing behavior though when using run stages.

File: perfkitbenchmarker/benchmark_spec.py (hunk 1 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: done

File: perfkitbenchmarker/vm_util.py (hunk 3 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: I'll do that on Monday - nit is done

File: perfkitbenchmarker/vm_util.py (hunk 4 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: done

File: perfkitbenchmarker/vm_util.py (hunk 5 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: done

File: perfkitbenchmarker/pkb.py (hunk 6 above)
Comment (ehankland) [inline perfkitbenchmarker/pkb.py]: done

File: perfkitbenchmarker/vm_util.py (hunk 7 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: done

File: perfkitbenchmarker/vm_util.py (hunk 8 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: done

**Comment by ehankland** [comment]
Still need to move the FlagValuesProxy class (I'll do it on Monday), otherwise fixed everything - thanks for the review - looks much cleaner now

File: README.md [hunk 9]
```diff
@@ -661,6 +661,29 @@ iperf:
...
+top level in a benchmark config. The expected value is a dictionary mapping
+flag names to their new default values. The flags are only defaults; it's still
+possible to override them via the command line. It's even possible to specify
+conflicting values of the same flag in different benchmarks:
+
+```
+iperf:
+  flags:
+    machine_type: n1-standard-2
+    zone: us-central1-b
```
Comment (cmccoy) [inline README.md]: `zones`?

Comment (ehankland) [inline README.md]: `zone` is the correct flag now after the config PR. If you think we need to keep the `zones` flag, I can put it back in, but it feels like kind of a hack - especially now that there's a better way (assigning zones to vm groups rather than individual vms)

Comment (cmccoy) [inline README.md]: No need, I didn't realize it had changed.

**Comment by ehankland** [comment]
PTAL

File: perfkitbenchmarker/context_util.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2015 Google Inc. All rights reserved.
```
Comment (cmccoy) [inline perfkitbenchmarker/context_util.py]: How about just `context`?

File: perfkitbenchmarker/data/__init__.py [hunk 11]
```diff
@@ -28,9 +28,10 @@
 
 import pkg_resources
 
+from perfkitbenchmarker import context_util
 from perfkitbenchmarker import flags
 
-FLAGS = flags.FLAGS
+FLAGS = context_util.FLAGS
```
Comment (cmccoy) [inline perfkitbenchmarker/data/__init__.py]: I'm a little concerned about references to `flags.FLAGS` creeping in elsewhere. Is there any way to make a single `FLAGS` import location?

Comment (cmccoy) [inline perfkitbenchmarker/data/__init__.py]: (`context_util.FLAGS` being preferred to `flags.FLAGS` is a bit confusing)

File: perfkitbenchmarker/context_util.py (hunk 10 above)
Comment (ehankland) [inline perfkitbenchmarker/context_util.py]: Done

File: perfkitbenchmarker/data/__init__.py (hunk 11 above)
Comment (ehankland) [inline perfkitbenchmarker/data/__init__.py]: I see your point - do you have any ideas?

Comment (cmccoy) [inline perfkitbenchmarker/data/__init__.py]: No, I don't think I have anything good. Attaching flags to the BenchmarkSpec *and* changing references from module-level flags to benchmark spec flags instances would help, but would require passing flag values through to packages, etc.

Comment (ehankland) [inline perfkitbenchmarker/data/__init__.py]: I added a test which should catch any new `FLAGS = flags.FLAGS`

**Comment by ehankland** [comment]
I tried something different (replaced flags.FLAGS with the flag proxy) - can you PTAL?

File: perfkitbenchmarker/context.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+# limitations under the License.
+
+"""Module for working with the current thread context."""
+
+import threading
+
+import gflags as flags
+
+
+class FlagsModuleProxy(object):
```
Comment (cmccoy) [inline perfkitbenchmarker/context.py]: Could you add a docstring?

File: perfkitbenchmarker/__init__.py [hunk 13]
```diff
@@ -12,5 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import gflags as flags  # NOQA
 import gflags_validators as flags_validators  # NOQA
+
+from perfkitbenchmarker import context
+
+
+flags = context.FlagsModuleProxy()
```
Comment (cmccoy) [inline perfkitbenchmarker/__init__.py]: Could you add a short description of the purpose of this?

File: perfkitbenchmarker/context.py (hunk 12 above)
Comment (cmccoy) [inline perfkitbenchmarker/context.py]: And a short test would be nice :)

**Comment by cmccoy** [comment]
I like the latest. A few documentation / testing nits, otherwise LGTM.

File: perfkitbenchmarker/context.py (hunk 12 above)
Comment (ehankland) [inline perfkitbenchmarker/context.py]: Done

File: perfkitbenchmarker/__init__.py (hunk 13 above)
Comment (ehankland) [inline perfkitbenchmarker/__init__.py]: Done

**Comment by ehankland** [comment]
PTAL

**Comment by cmccoy** [comment]
LGTM. Two more nits, sorry. Feel free to merge after that.
